alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
+        let popUp = NSPopUpButton()
+        popUp.addItems(withTitles: values)
+        popUp.selectItem(withTitle: Preferences.rawValues[rawName]!)
+
+        self[keyPath: keyPath] = popUp
+
+        return makeLabelWithProvidedControl(labelText, rawName, popUp, suffixText)
+    }
+
+    private func makeLabelWithSlider(_ keyPath: ReferenceWritableKeyPath<PreferencesPanel, NSSlider?>, _ labelText: String, rawName: String, minValue: Double, maxValue: Double, numberOfTickMarks: Int, suffixText: String? = nil) -> NSStackView {
+        let slider = NSSlider(
+                value: Double(Preferences.rawValues[rawName]!)!,
+                minValue: minValue,
+                maxValue: maxValue,
+                target: self,
+                action: #selector(controlDidEndEditing)
+        )
+        slider.numberOfTickMarks = numberOfTickMarks
+        slider.allowsTickMarkValuesOnly = numberOfTickMarks > 1
+        slider.tickMarkPosition = .below
+
+        self[keyPath: keyPath] = slider
+
+        return makeLabelWithProvidedControl(labelText, rawName, slider, suffixText)
+    }
+
+    private func makeLabelWithProvidedControl(_ labelText: String, _ rawName: String, _ control: NSControl, _ suffixText: String? = nil) -> NSStackView {
+        let label = NSTextField(wrappingLabelWithString: labelText + ": ")
+        label.alignment = .right
+        label.widthAnchor.constraint(equalToConstant: calcLabelWidth()).isActive = true
+
+        control.identifier = NSUserInterfaceItemIdentifier(rawName)
+        control.target = self
+        control.action = #selector(controlDidEndEditing)
+        let containerView = NSStackView(views: [label, control])
+
+        if suffixText != nil {
+            let suffix = NSTextField(labelWithString: suffixText!)
+            suffix.textColor = .gray
+            containerView.addView(suffix, in: .leading)
         }
+
+        return containerView
     }
 
-    func textDidEndEditing(_ notification: Notification) {
-        if let textView = notification.object as? NSTextView {
-            let key = inputsMap[textView]!
+    /*
+    usage notes:
+    - NSSlider: supports on purpose currently only decimal values
+    */
+    @objc func controlDidEndEditing(senderControl: NSControl) {
+        self.makeFirstResponder(invisibleTextField) // deselects any possibly selected NSTextField (so slider & popUp changes deselect them)
+
+        let key: String? = senderControl.identifier?.rawValue
+        var newValue: String?
+
+        if senderControl is NSPopUpButton {
+            newValue = (senderControl as! NSPopUpButton).titleOfSelectedItem!
+        } else if senderControl is NSSlider {
+            newValue = String(format: "%.0f", Double(senderControl.stringValue)!) // we are only interested in decimals of the provided double
```

Comment:
That part is confusing. Is it some kind of trick where the value is like `0.80` and you want to extract `80`. I hope you can either clarify what the trick is about and why it's necessary, and simply do without this when you iterate on the sliders to give them text marks

Reply:
NSSlider works internally with Doubles. So if the user selects a value it could be either 78.875 or 80 (depending on if NSSlider.allowsTickMarkValuesOnly is set). Our Preferences implementation expects decimals (since the internalisation thing with 0.8 and 0,8 I reported in #51 and you changed the maxScreenUsage from 0.8 to be 80). This code snippet was one of a few proposals on stackoverflow to get a decimal from a double value in Swift. Thats the one I picked.

The NSSlider Text Marks will probably be just some representation of the internal Double value and not be used as its actual value.

Reply:
> Btw. I was unhappy with the default macOS altTab experience and had on my todo list to look into alternatives but an issue was their closed source nature. I just can't install that on my company MacBook without hesitation or approval (security, privacy, trust). Then Ive read about your application in an [article](https://t3n.de/news/macos-windows-fensterwechsel-alt-tab-1209176/) which was presented one day in my google news feed (I had googled for all the alternatives 😄 weeks or months ago). Thanks for this project/application!

What an uplifting story! Thanks for sharing! :)

Reply:
You'r welcome 🥇